tsconvert/newick.py

```diff
@@ -123,23 +124,72 @@ def to_ms(ts):
     return output
 
 
-def from_newick(string):
+def from_newick(string, min_edge_length=0):
     """
     Returns a tree sequence representation of the specified newick string.
+
+    The tree sequence will contain a single tree, as specified by the newick. All
+    leaf nodes will be marked as samples (``tskit.NODE_IS_SAMPLE``).
+
+    :param string string: Newick string
+    :param float min_edge_length: Replace any edge length shorter than this value by this
+        value. Unlike newick, tskit doesn't support zero or negative edge lengths, so
+        setting this argument to a small value is necessary when importing trees with
+        zero or negative lengths.
     """
-    tree = dendropy.Tree.get(data=string, schema="newick")
+    trees = newick.loads(string)
+    if len(trees) > 1:
+        raise ValueError("Only one tree can be imported from a newick string")
+    if len(trees) == 0:
+        raise ValueError("Newick string was empty")
+    tree = trees[0]
     tables = tskit.TableCollection(1)
+    tables.nodes.metadata_schema = tskit.MetadataSchema(
+        {
+            "codec": "json",
+            "type": "object",
+            "properties": {
+                "newick_id": {
+                    "type": ["string", "null"],
+                    "description": "ID from newick file",
+                }
+            },
+        }
+    )
 
     id_map = {}
-    for node in tree.ageorder_node_iter():
-        children = list(node.child_nodes())
+
+    def get_or_add_node(node, time):
         if node not in id_map:
-            flags = tskit.NODE_IS_SAMPLE if len(children) == 0 else 0
-            # TODO derive information from the node and store it as JSON metadata.
-            id_map[node] = tables.nodes.add_row(flags=flags, time=node.age)
-        node_id = id_map[node]
-        for child in children:
-            tables.edges.add_row(0, 1, node_id, id_map[child])
+            flags = tskit.NODE_IS_SAMPLE if len(node.descendants) == 0 else 0
+            id_map[node] = (
```

Comment:
Maybe it'd be a bit more elegant if we did this:
```python
tables.nodes.add_row(...)
id_map[node] = tables.nodes[-1]
```
It would avoid needing to package up tuples (which is nearly always bad)

Reply:
Yeah, I've ditched the tuple and just look up the node.